File: tools/fwmod/fwmod/format.py
"""The .fwmod container codec (writer + reader).

This is the Python half of a format defined once in docs/fwmod-format.md and
implemented twice (here, and in TS at packages/emulator-core/src/fwmod/). A
committed golden fixture is decoded by both sides' tests so they cannot drift.

Stdlib only: struct for the header, zlib for CRC-32.
"""

from __future__ import annotations

import struct
import zlib
from dataclasses import dataclass, field
from enum import IntEnum

MAGIC = b"FWMD"
FORMAT_VERSION = 1
ABI_VERSION = 1
HEADER_SIZE = 32

# Header layout (little-endian): magic[4], format_version u16, abi_version u16,
# arch u8, flags u8, reserved u16, code_size u32, bss_size u32, entry_offset u32,
# load_addr u32, crc32 u32. See docs/fwmod-format.md for the field table.
_HEADER = struct.Struct("<4sHHBBHIIIII")
assert _HEADER.size == HEADER_SIZE

FLAG_PIC = 0x01


class Arch(IntEnum):
    UNKNOWN = 0
    XTENSA_LX7 = 1
    HOST_X86 = 2
    HOST_X86_64 = 3
    WASM32 = 4

    @classmethod
    def from_name(cls, name: str) -> "Arch":
        try:
            return _ARCH_BY_NAME[name]
        except KeyError:
            valid = ", ".join(_ARCH_BY_NAME)
            raise ValueError(f"unknown arch {name!r} (expected one of: {valid})")

    @property
    def cli_name(self) -> str:
        return _ARCH_TO_NAME[self]


_ARCH_BY_NAME = {
    "unknown": Arch.UNKNOWN,
    "xtensa-lx7": Arch.XTENSA_LX7,
    "host-x86": Arch.HOST_X86,
    "host-x86_64": Arch.HOST_X86_64,
    "wasm32": Arch.WASM32,
}
_ARCH_TO_NAME = {v: k for k, v in _ARCH_BY_NAME.items()}


def crc32(payload: bytes) -> int:
    """CRC-32/ISO-HDLC (zlib variant). crc32(b"123456789") == 0xCBF43926."""
    return zlib.crc32(payload) & 0xFFFFFFFF


@dataclass
class FwModule:
    """A decoded .fwmod: its header fields plus the raw payload."""

    payload: bytes
    arch: Arch = Arch.UNKNOWN
    abi_version: int = ABI_VERSION
    format_version: int = FORMAT_VERSION
    flags: int = 0
    bss_size: int = 0
    entry_offset: int = 0
    load_addr: int = 0
    reserved: int = 0
    # Populated by decode(); the CRC stored in the file (may differ from the
    # recomputed CRC for a corrupt file). encode() always writes a fresh CRC.
    stored_crc32: int | None = field(default=None)
    # Populated by decode() with the total input length, so validate() can flag
    # trailing bytes after the payload. None for constructor-built modules.
    source_size: int | None = field(default=None)

    @property
    def code_size(self) -> int:
        return len(self.payload)

    @property
    def computed_crc32(self) -> int:
        return crc32(self.payload)

    def _field_ranges(self) -> tuple[tuple[str, int, int], ...]:
        """(name, value, max) for every fixed-width header field."""
        return (
            ("format_version", self.format_version, 0xFFFF),
            ("abi_version", self.abi_version, 0xFFFF),
            ("reserved", self.reserved, 0xFFFF),
            ("arch", int(self.arch), 0xFF),
            ("flags", self.flags, 0xFF),
            ("code_size", self.code_size, 0xFFFFFFFF),
            ("bss_size", self.bss_size, 0xFFFFFFFF),
            ("entry_offset", self.entry_offset, 0xFFFFFFFF),
            ("load_addr", self.load_addr, 0xFFFFFFFF),
        )

    def encode(self) -> bytes:
        """Serialize to bytes, computing the payload CRC fresh. Raises ValueError
        on an out-of-range field rather than letting struct.pack raise."""
        for name, value, hi in self._field_ranges():
            if not (0 <= value <= hi):
                raise ValueError(f"{name} out of range: {value} (allowed 0..{hi})")
        header = _HEADER.pack(
            MAGIC,
            self.format_version,
            self.abi_version,
            int(self.arch),
            self.flags,
            self.reserved,
            self.code_size,
            self.bss_size,
            self.entry_offset,
            self.load_addr,
            self.computed_crc32,
        )
        return header + self.payload

    def validate(self) -> list[str]:
        """Return a list of problems; empty means a loadable module."""
        problems: list[str] = []
        for name, value, hi in self._field_ranges():
            if not (0 <= value <= hi):
                problems.append(f"{name} out of range: {value} (allowed 0..{hi})")
        if self.format_version != FORMAT_VERSION:
            problems.append(
                f"unsupported format_version {self.format_version} "
                f"(this tool writes {FORMAT_VERSION})"
            )
        if self.reserved != 0:
            problems.append(f"reserved field must be 0, got {self.reserved}")
        if self.code_size < 1:
            problems.append("empty payload (code_size must be >= 1)")
        if self.entry_offset >= self.code_size:
            problems.append(
                f"entry_offset {self.entry_offset} is outside the "
                f"{self.code_size}-byte payload"
            )
        if self.stored_crc32 is not None and self.stored_crc32 != self.computed_crc32:
            problems.append(
                f"crc32 mismatch: header says {self.stored_crc32:#010x}, "
                f"payload is {self.computed_crc32:#010x}"
            )
        if (
            self.source_size is not None
            and self.source_size != HEADER_SIZE + self.code_size
        ):
            trailing = self.source_size - HEADER_SIZE - self.code_size
            problems.append(
                f"file is {self.source_size} bytes, expected {HEADER_SIZE + self.code_size} "
                f"({trailing:+d} trailing byte(s) after the payload)"
            )
        return problems


def decode(data: bytes) -> FwModule:
    """Parse a .fwmod buffer. Lenient: parses any 32+ byte buffer with the right
    magic so tooling can inspect malformed files. Use validate() to gate loading.
    """
    if len(data) < HEADER_SIZE:
        raise ValueError(
            f"too small to be a .fwmod: {len(data)} bytes (need >= {HEADER_SIZE})"
        )
    (
        magic,
        format_version,
        abi_version,
        arch,
        flags,
        reserved,
        code_size,
        bss_size,
        entry_offset,
        load_addr,
        stored_crc,
    ) = _HEADER.unpack_from(data, 0)
    if magic != MAGIC:
        raise ValueError(f"bad magic {magic!r} (expected {MAGIC!r})")

    available = len(data) - HEADER_SIZE
    if code_size > available:
        raise ValueError(
            f"truncated: header claims {code_size}-byte payload but only "
            f"{available} bytes follow the header"
        )
    payload = data[HEADER_SIZE : HEADER_SIZE + code_size]

    return FwModule(
        payload=payload,
        arch=Arch(arch) if arch in _ARCH_TO_NAME else Arch.UNKNOWN,
        abi_version=abi_version,
        format_version=format_version,
        flags=flags,
        bss_size=bss_size,
        entry_offset=entry_offset,
        load_addr=load_addr,
        reserved=reserved,
        stored_crc32=stored_crc,
        source_size=len(data),
    )

File: packages/emulator-core/src/fwmod/format.ts
/**
 * The .fwmod container codec (reader + writer).
 *
 * This is the TypeScript half of a format defined once in docs/fwmod-format.md
 * and implemented twice (here, and in Python at tools/fwmod/fwmod/format.py). A
 * committed golden fixture is decoded by both sides' tests so they cannot drift.
 *
 * The emulator uses this to load, validate, and inspect native modules.
 * Executing them is Phase 5; here we parse and integrity-check the container.
 */

import { crc32 } from "./crc32.js";

export const MAGIC = "FWMD";
export const FORMAT_VERSION = 1;
export const ABI_VERSION = 1;
export const HEADER_SIZE = 32;

export const FLAG_PIC = 0x01;

/** Target architecture (`arch` byte). Host/wasm are dev targets. */
export const Arch = {
  Unknown: 0,
  XtensaLx7: 1, // ESP32-S3 — the real hardware target
  HostX86: 2,
  HostX86_64: 3,
  Wasm32: 4,
} as const;
export type Arch = (typeof Arch)[keyof typeof Arch];

const ARCH_NAMES: Readonly<Record<number, string>> = {
  0: "unknown",
  1: "xtensa-lx7",
  2: "host-x86",
  3: "host-x86_64",
  4: "wasm32",
};

export function archName(arch: number): string {
  return ARCH_NAMES[arch] ?? `arch#${arch}`;
}

export interface FwModuleFields {
  payload: Uint8Array;
  arch?: number;
  abiVersion?: number;
  formatVersion?: number;
  flags?: number;
  bssSize?: number;
  entryOffset?: number;
  loadAddr?: number;
  reserved?: number;
  /** CRC stored in the file. Set by decode(); may differ from the recomputed
   *  CRC for a corrupt file. encode() always writes a fresh CRC. */
  storedCrc32?: number | null;
  /** Total input length, set by decode() so validate() can flag trailing bytes.
   *  null for constructor-built modules. */
  sourceSize?: number | null;
}

/** A decoded .fwmod: its header fields plus the raw payload. */
export class FwModule {
  readonly payload: Uint8Array;
  readonly arch: number;
  readonly abiVersion: number;
  readonly formatVersion: number;
  readonly flags: number;
  readonly bssSize: number;
  readonly entryOffset: number;
  readonly loadAddr: number;
  readonly reserved: number;
  readonly storedCrc32: number | null;
  readonly sourceSize: number | null;

  constructor(fields: FwModuleFields) {
    this.payload = fields.payload;
    this.arch = fields.arch ?? Arch.Unknown;
    this.abiVersion = fields.abiVersion ?? ABI_VERSION;
    this.formatVersion = fields.formatVersion ?? FORMAT_VERSION;
    this.flags = fields.flags ?? 0;
    this.bssSize = fields.bssSize ?? 0;
    this.entryOffset = fields.entryOffset ?? 0;
    this.loadAddr = fields.loadAddr ?? 0;
    this.reserved = fields.reserved ?? 0;
    this.storedCrc32 = fields.storedCrc32 ?? null;
    this.sourceSize = fields.sourceSize ?? null;
  }

  /** (name, value, max) for every fixed-width header field. */
  private fieldRanges(): ReadonlyArray<readonly [string, number, number]> {
    return [
      ["format_version", this.formatVersion, 0xffff],
      ["abi_version", this.abiVersion, 0xffff],
      ["reserved", this.reserved, 0xffff],
      ["arch", this.arch, 0xff],
      ["flags", this.flags, 0xff],
      ["code_size", this.codeSize, 0xffffffff],
      ["bss_size", this.bssSize, 0xffffffff],
      ["entry_offset", this.entryOffset, 0xffffffff],
      ["load_addr", this.loadAddr, 0xffffffff],
    ];
  }

  get codeSize(): number {
    return this.payload.length;
  }

  get computedCrc32(): number {
    return crc32(this.payload);
  }

  get archLabel(): string {
    return archName(this.arch);
  }

  /** Serialize to bytes, computing the payload CRC fresh. Throws on an
   *  out-of-range field rather than silently truncating it (matches the
   *  Python writer, which raises). */
  encode(): Uint8Array {
    for (const [name, value, hi] of this.fieldRanges()) {
      if (!Number.isInteger(value) || value < 0 || value > hi) {
        throw new Error(`${name} out of range: ${value} (allowed 0..${hi})`);
      }
    }
    const out = new Uint8Array(HEADER_SIZE + this.codeSize);
    const view = new DataView(out.buffer);
    for (let i = 0; i < 4; i++) out[i] = MAGIC.charCodeAt(i);
    view.setUint16(4, this.formatVersion, true);
    view.setUint16(6, this.abiVersion, true);
    out[8] = this.arch;
    out[9] = this.flags;
    view.setUint16(10, this.reserved, true);
    view.setUint32(12, this.codeSize, true);
    view.setUint32(16, this.bssSize, true);
    view.setUint32(20, this.entryOffset, true);
    view.setUint32(24, this.loadAddr, true);
    view.setUint32(28, this.computedCrc32, true);
    out.set(this.payload, HEADER_SIZE);
    return out;
  }

  /** Return a list of problems; an empty list means a loadable module. */
  validate(): string[] {
    const problems: string[] = [];
    for (const [name, value, hi] of this.fieldRanges()) {
      if (!Number.isInteger(value) || value < 0 || value > hi) {
        problems.push(`${name} out of range: ${value} (allowed 0..${hi})`);
      }
    }
    if (this.formatVersion !== FORMAT_VERSION) {
      problems.push(
        `unsupported format_version ${this.formatVersion} (this build reads ${FORMAT_VERSION})`,
      );
    }
    if (this.reserved !== 0) {
      problems.push(`reserved field must be 0, got ${this.reserved}`);
    }
    if (this.codeSize < 1) {
      problems.push("empty payload (code_size must be >= 1)");
    }
    if (this.entryOffset >= this.codeSize) {
      problems.push(
        `entry_offset ${this.entryOffset} is outside the ${this.codeSize}-byte payload`,
      );
    }
    if (this.storedCrc32 !== null && this.storedCrc32 !== this.computedCrc32) {
      problems.push(
        `crc32 mismatch: header says ${hex(this.storedCrc32)}, payload is ${hex(this.computedCrc32)}`,
      );
    }
    if (
      this.sourceSize !== null &&
      this.sourceSize !== HEADER_SIZE + this.codeSize
    ) {
      const trailing = this.sourceSize - HEADER_SIZE - this.codeSize;
      problems.push(
        `file is ${this.sourceSize} bytes, expected ${HEADER_SIZE + this.codeSize} (${trailing > 0 ? "+" : ""}${trailing} trailing byte(s) after the payload)`,
      );
    }
    return problems;
  }
}

/**
 * Parse a .fwmod buffer. Lenient: parses any 32+ byte buffer with the right
 * magic so tooling can inspect malformed files. Throws on a structurally
 * impossible buffer (too small, bad magic, truncated payload). Use validate()
 * to gate loading.
 */
export function decodeFwmod(data: Uint8Array): FwModule {
  if (data.length < HEADER_SIZE) {
    throw new Error(
      `too small to be a .fwmod: ${data.length} bytes (need >= ${HEADER_SIZE})`,
    );
  }
  // DataView over the exact region the array views (handles subarray offsets).
  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);

  let magic = "";
  for (let i = 0; i < 4; i++) magic += String.fromCharCode(data[i]);
  if (magic !== MAGIC) {
    throw new Error(`bad magic ${JSON.stringify(magic)} (expected "${MAGIC}")`);
  }

  const formatVersion = view.getUint16(4, true);
  const abiVersion = view.getUint16(6, true);
  // Coerce an unrecognized arch byte to Unknown, matching the Python decoder.
  const rawArch = data[8];
  const arch = rawArch in ARCH_NAMES ? rawArch : Arch.Unknown;
  const flags = data[9];
  const reserved = view.getUint16(10, true);
  const codeSize = view.getUint32(12, true);
  const bssSize = view.getUint32(16, true);
  const entryOffset = view.getUint32(20, true);
  const loadAddr = view.getUint32(24, true);
  const storedCrc32 = view.getUint32(28, true);

  const available = data.length - HEADER_SIZE;
  if (codeSize > available) {
    throw new Error(
      `truncated: header claims ${codeSize}-byte payload but only ${available} bytes follow the header`,
    );
  }
  const payload = data.subarray(HEADER_SIZE, HEADER_SIZE + codeSize);

  return new FwModule({
    payload,
    arch,
    abiVersion,
    formatVersion,
    flags,
    bssSize,
    entryOffset,
    loadAddr,
    reserved,
    storedCrc32,
    sourceSize: data.length,
  });
}

function hex(n: number): string {
  return `0x${(n >>> 0).toString(16).padStart(8, "0")}`;
}
